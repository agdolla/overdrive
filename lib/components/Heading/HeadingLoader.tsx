import cx from 'clsx';
import React, { FunctionComponent, memo } from 'react';
import { LoadingBox } from '../LoadingBox';
import styles from './style.scss';
import { ESize as EHeadingSize } from './Heading';

export interface IProps {
	size?: EHeadingSize;
	className?: string;
	randomWidth?: boolean;
	blinking?: boolean;
}

const cssSizeMap: Map<EHeadingSize, string> = new Map([
	[EHeadingSize.Heading1, styles.loaderHeading1],
	[EHeadingSize.Heading2, styles.loaderHeading2],
	[EHeadingSize.Heading3, styles.loaderHeading3],
	[EHeadingSize.Heading4, styles.loaderHeading4],
	[EHeadingSize.Heading5, styles.loaderHeading5],
	[EHeadingSize.Heading6, styles.loaderHeading6],
]);

const HeadingLoaderComponent: FunctionComponent<IProps> = ({
	size = EHeadingSize.Heading1,
	randomWidth = false,
	blinking = true,
	className = '',
}) => (
	<LoadingBox
		blinking={blinking}
		className={cx([styles.root, cssSizeMap.get(size), className])}
		randomWidth={randomWidth}
	/>
);

export const HeadingLoader = memo(HeadingLoaderComponent);
