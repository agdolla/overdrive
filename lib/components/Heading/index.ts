export {
	Heading,
	ESize as EHeadingSize,
	EVariant as EHeadingVariant,
	IProps as IHeadingProps,
} from './Heading';
export { HeadingLoader, IProps as IHeadingLoaderProps } from './HeadingLoader';
