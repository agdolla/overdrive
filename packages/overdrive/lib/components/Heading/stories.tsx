import * as React from 'react';
import { ComponentProps } from 'react';

import { Heading } from '.';
import { ArgTypes } from '@storybook/react';

const headingTypeOptions: Array<ComponentProps<typeof Heading>['is']> = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'];
const sizeScale: Array<ComponentProps<typeof Heading>['size']> = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];
const alignOptions: Array<ComponentProps<typeof Heading>['align']> = ['left', 'center', 'right'];
const colourOptions: Array<ComponentProps<typeof Heading>['colour']> = ['dark', 'light', 'neutral', 'primary', 'secondary', 'white', 'information', 'link', 'success', 'danger', 'warning', 'shine'];

export default {
	title: 'Foundation/Typography/Heading',
	component: Heading,
};


const Template = (args) => (
	<div style={{ maxWidth: '350px', width: '100%' }}><Heading {...args} /></div>
);

const AllTypesTemplate = (args) => (
	<div style={{ maxWidth: '350px', width: '100%' }}>
		{headingTypeOptions.map(is => (
			<Heading key={is} {...args} is={is} />
		))}
	</div>
);

const AllColoursTemplate = (args) => (
	<div style={{ maxWidth: '350px', width: '100%' }}>
		{colourOptions.map(colour => (
			<div style={{marginBottom: 8}}>
				{headingTypeOptions.map((is, index)=> (
					<Heading key={`${is}-${index}`} {...args} colour={colour} is={is} />
				))}
			</div>
		))}
	</div>
);

const sharedArgTypes: ArgTypes = {
	size: {
		options: sizeScale,
		defaultValue: void 0,
		control: {
			type: 'select',
		},
	},
	align: {
		options: alignOptions,
		defaultValue: 'left',
		control: {
			type: 'select',
		},
	},
};
const colourArgTypes: ArgTypes = {
	colour: {
		options: colourOptions,
		defaultValue: void 0,
		control: {
			type: 'select',
		},
	},
};
const argTypes: ArgTypes = {
	...sharedArgTypes,
	...colourArgTypes,
	is: {
		options: headingTypeOptions,
		defaultValue: 'h1',
		table: {
			type: { summary: 'heading' },
			defaultValue: 'h1',
		},
		description: 'HTML heading tag to be used',
		control: {
			type: 'select',
		},
	},
};

const standardProps: ComponentProps<typeof Heading> = {
	is: 'h1',
	children: 'I am a heading',
};
const allTypesProps: ComponentProps<typeof Heading> = {
	children: 'I am a heading',
};
export const standard = Template.bind(standardProps);
standard.args = standardProps;
standard.argTypes = argTypes;

export const allTypes = AllTypesTemplate.bind(allTypesProps);
allTypes.args = allTypesProps;
allTypes.argTypes = { ...sharedArgTypes, colourArgTypes };

export const allColours = AllColoursTemplate.bind(allTypesProps);
allColours.args = allTypesProps;
allColours.argTypes = sharedArgTypes;
