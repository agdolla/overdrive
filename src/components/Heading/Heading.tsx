import cx from 'clsx';
import React, { FunctionComponent } from 'react';
import styles from './style.scss';

export interface IProps {
	variant?: EVariant | null;
	size?: ESize;
	className?: string;
}

export enum EVariant {
	Mobile = 'mobile',
	Desktop = 'desktop',
}

export enum ESize {
	Heading1 = 'heading-1',
	Heading2 = 'heading-2',
	Heading3 = 'heading-3',
	Heading4 = 'heading-4',
	Heading5 = 'heading-5',
	Heading6 = 'heading-6',
}

const cssSizeMap: Map<ESize, string> = new Map([
	[ESize.Heading1, styles.Heading1],
	[ESize.Heading2, styles.Heading2],
	[ESize.Heading3, styles.Heading3],
	[ESize.Heading4, styles.Heading4],
	[ESize.Heading5, styles.Heading5],
	[ESize.Heading6, styles.Heading6],
]);

const headingTagsMap = new Map<ESize, string>([
	[ESize.Heading1, 'h1'],
	[ESize.Heading2, 'h2'],
	[ESize.Heading3, 'h3'],
	[ESize.Heading4, 'h4'],
	[ESize.Heading5, 'h5'],
	[ESize.Heading6, 'h6'],
]);

export const Heading: FunctionComponent<IProps> = ({
	variant = null,
	size = ESize.Heading1,
	className = '',
	children,
}) => {
	const cls = cx([styles.root, cssSizeMap.get(size), className], {
		[styles.variantMobile]: !!variant && variant === EVariant.Mobile,
		[styles.variantDesktop]: !!variant && variant === EVariant.Desktop,
	});

	const Tag = headingTagsMap.get(size) as any; // TODO: Fix this any

	return <Tag className={cls} size={size} children={children} />;
};
