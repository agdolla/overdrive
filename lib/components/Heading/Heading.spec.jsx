import React from 'react';
import { mount, render, shallow } from 'enzyme';
import { Heading } from './Heading';
import { EHeadingSize, EHeadingVariant } from './index';
import { HeadingLoader } from './HeadingLoader';

const testLabel = 'Hello World!';
describe('<Heading />', () => {
	it('should not throw', () =>
		expect(() => shallow(<Heading />)).not.toThrow());

	it('should pass on className to dom element', () => {
		const heading = mount(<Heading className="heading-class" />);
		expect(heading.hasClass('heading-class')).toBeTruthy();
		heading.unmount();
	});

	it('should match snapshot for default heading', () => {
		expect(shallow(<Heading>{testLabel}</Heading>)).toMatchSnapshot();
	});

	it('should match snapshot for default mobile heading', () => {
		expect(
			shallow(
				<Heading variant={EHeadingVariant.Mobile}>{testLabel}</Heading>
			)
		).toMatchSnapshot();
	});

	it('should apply class for mobile variant', () => {
		const heading = shallow(
			<Heading variant={EHeadingVariant.Mobile}>{testLabel}</Heading>
		);
		expect(heading.hasClass('variantMobile')).toBeTruthy();
	});

	it('should apply class for desktop variant', () => {
		const heading = shallow(
			<Heading variant={EHeadingVariant.Desktop}>{testLabel}</Heading>
		);
		expect(heading.hasClass('variantDesktop')).toBeTruthy();
	});

	it('should pass the child down to DOM element', () => {
		expect(shallow(<Heading>{testLabel}</Heading>).text()).toEqual(
			testLabel
		);
	});

	it('should match snapshot for default heading 1', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading1} />)
		).toMatchSnapshot();
	});

	it('should match snapshot for default heading 2', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading2} />)
		).toMatchSnapshot();
	});

	it('should match snapshot for default heading 3', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading3} />)
		).toMatchSnapshot();
	});

	it('should match snapshot for default heading 4', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading4} />)
		).toMatchSnapshot();
	});

	it('should match snapshot for default heading 5', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading5} />)
		).toMatchSnapshot();
	});

	it('should match snapshot for default heading 6', () => {
		expect(
			shallow(<Heading size={EHeadingSize.Heading6} />)
		).toMatchSnapshot();
	});

	it('should use h1 element for heading 1', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading1} />);
		expect(heading.type()).toEqual(`h1`);
	});

	it('should use h2 element for heading 2', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading2} />);
		expect(heading.type()).toEqual(`h2`);
	});

	it('should use h3 element for heading 3', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading3} />);
		expect(heading.type()).toEqual(`h3`);
	});

	it('should use h4 element for heading 4', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading4} />);
		expect(heading.type()).toEqual(`h4`);
	});

	it('should use h5 element for heading 5', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading5} />);
		expect(heading.type()).toEqual(`h5`);
	});

	it('should use h6 element for heading 6', () => {
		const heading = shallow(<Heading size={EHeadingSize.Heading6} />);
		expect(heading.type()).toEqual(`h6`);
	});

	describe('when loading', function() {
		it('should match snapshot for loading view', () => {
			expect(
				shallow(<HeadingLoader variant={EHeadingVariant.Desktop} />)
			).toMatchSnapshot();
		});

		it('should add a loading box', () => {
			expect(
				shallow(
					<HeadingLoader
						size={EHeadingSize.Heading1}
						variant={EHeadingVariant.Desktop}
					/>
				).hasClass('loaderHeading1')
			).toEqual(true);
		});

		it('should not add random width style when randomWidth is false', () => {
			expect(
				render(
					<HeadingLoader
						size={EHeadingSize.Heading2}
						randomWidth={false}
						variant={EHeadingVariant.Desktop}
					/>
				).prop('style').length
			).toBeFalsy();
		});

		it('should add random width style when randomWidth is true', () => {
			expect(
				render(
					<HeadingLoader
						size={EHeadingSize.Heading2}
						randomWidth={true}
						variant={EHeadingVariant.Desktop}
					/>
				)
					.prop('style')
					.width.includes('%')
			).toBeTruthy();
		});
	});
});
