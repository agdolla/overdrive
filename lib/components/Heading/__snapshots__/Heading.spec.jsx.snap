// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should match snapshot for default heading 1`] = `
<h1
  className="root Heading1"
  size="heading-1"
>
  Hello World!
</h1>
`;

exports[`<Heading /> should match snapshot for default heading 1 1`] = `
<h1
  className="root Heading1"
  size="heading-1"
/>
`;

exports[`<Heading /> should match snapshot for default heading 2 1`] = `
<h2
  className="root Heading2"
  size="heading-2"
/>
`;

exports[`<Heading /> should match snapshot for default heading 3 1`] = `
<h3
  className="root Heading3"
  size="heading-3"
/>
`;

exports[`<Heading /> should match snapshot for default heading 4 1`] = `
<h4
  className="root Heading4"
  size="heading-4"
/>
`;

exports[`<Heading /> should match snapshot for default heading 5 1`] = `
<h5
  className="root Heading5"
  size="heading-5"
/>
`;

exports[`<Heading /> should match snapshot for default heading 6 1`] = `
<h6
  className="root Heading6"
  size="heading-6"
/>
`;

exports[`<Heading /> should match snapshot for default mobile heading 1`] = `
<h1
  className="root Heading1 variantMobile"
  size="heading-1"
>
  Hello World!
</h1>
`;

exports[`<Heading /> when loading should match snapshot for loading view 1`] = `
<LoadingBoxComponent
  blinking={true}
  className="root loaderHeading1"
  randomWidth={false}
/>
`;
